docs: sort entries in contributors.json alphabetically

AndrewKushnir opened the pull request:
Currently entries in the `contributors.json` file are not sorted and new entries are typically added at the very
end of the file, which is often causing merge conflicts. This commit updates the `contributors.json` to have all
entries sorted in alphabetical order (based on GitHub handles), which should reduce the number of merge conflicts.


## PR Type
What kind of change does this PR introduce?

- [x] Documentation content changes


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

mary-poppins commented:
You can preview 4ab6a44 at https://pr40369-4ab6a44.ngbuilds.io/.

gkalpak reviewed (commented):
Yesssss 💯 :grin:

I know this is still a draft (so not ready for review), but for the final PR it would be better to have one commit for sorting only and a different commit (or commits) for other changes (such as updating leads/mentors). 

BTW, currently we have a [script][1] to check for missing contributor images, which we call as part of [building AIO][2]. We could expand it to include more checks (such as alphabetically ordered keys :v:).

[1]: https://github.com/angular/angular/blob/6a9e3284329baf1bb6b621a8840ace629eb4b16e/aio/scripts/contributors/check-pictures.js
[2]: https://github.com/angular/angular/blob/6a9e3284329baf1bb6b621a8840ace629eb4b16e/aio/package.json#L56

AndrewKushnir committed:
docs: sort entries in contributors.json alphabetically

Currently entries in the `contributors.json` file are not sorted and new entries are typically added at the very
end of the file, which is often causing merge conflicts. This commit updates the `contributors.json` to have all
entries sorted in alphabetical order (based on GitHub handles), which should reduce the number of merge conflicts.

AndrewKushnir committed:
docs: include a note that all entries in contributors.json should be present in alphabetical order

AndrewKushnir committed:
refactor(docs-infra): rename contributors/check-pictures.js to validate-data.js

This commit renames contributors/check-pictures.js to contributors/validate-data.js, since this file script will also
be used to include other checks.

AndrewKushnir committed:
build(docs-infra): add more checks to verify contributors.json data

This commit adds extra steps to verify contributors.json data, such as checking `lead` and `mentor` field as well as
making sure the list is sorted alphabetically.

AndrewKushnir committed:
docs: resolve non-exiting references in contributors.json

The previous commit introduced extra checks to verify contributors.json data and this commit contains the necessary
fixes to the problems identified by the new set of checks.

AndrewKushnir commented:
@gkalpak, thanks for the feedback!

I've updated this PR to address your comments and also added extra logic to verify the content of the `contributors.json` file.

Could you please have another look when you get a chance?

Thank you.

mary-poppins commented:
You can preview dc70e96 at https://pr40369-dc70e96.ngbuilds.io/.

gkalpak commented on aio/package.json:
```diff
@@ -53,7 +53,7 @@
     "postcheck-env": "yarn aio-check-local",
     "payload-size": "scripts/payload.sh",
-    "predocs": "node scripts/contributors/check-pictures && yarn generate-stackblitz && yarn generate-zips",
+    "predocs": "node scripts/contributors/check-config && yarn generate-stackblitz && yarn generate-zips",
```
Nit: "config" sounds a little confusing to me (since it is not really a config file we are checking). Consider something `check-data` or `validate-data`.

gkalpak commented on aio/scripts/contributors/check-config.js:
```diff
@@ -0,0 +1,60 @@
+const CONTENT_DIR = resolve(__dirname, '../../content');
+const IMAGES_DIR = join(CONTENT_DIR, 'images/bios');
+const CONTRIBUTORS_PATH = join(CONTENT_DIR, 'marketing/contributors.json');
+const EXISTING_GROUPS = new Set(["Angular", "GDE", "Collaborators"]);
```
```suggestion
const EXISTING_GROUPS = new Set(['Angular', 'GDE', 'Collaborators']);
```

gkalpak commented on aio/scripts/contributors/check-config.js:
```diff
@@ -0,0 +1,60 @@
+      throw new Error(`The 'groups' field should be defined as a non-empty array (entry: '${key}').`);
+    }
+    if (entry.groups.some(group => !EXISTING_GROUPS.has(group))) {
+      throw new Error(`The '${key}' entry contains 'groups' field with some values outside of existing groups ` +
+        `(groups: ${JSON.stringify(entry.groups)}, existing groups: ${JSON.stringify(Array.from(EXISTING_GROUPS))}).`);
```
Nit: IMO, the message would become slightly clearer if we used "known" instead of "existing". E.g.:

```suggestion
      throw new Error(`The '${key}' entry contains 'groups' field with unknown values ` +
        `(groups: ${JSON.stringify(entry.groups)}, known values: ${JSON.stringify(Array.from(EXISTING_GROUPS))}).`);
```

gkalpak reviewed (approved):
A couple of minor comments/suggestions (nothing blocking). LGTM otherwise :+1:

Reviewed-for: global-docs-approvers

mary-poppins commented:
You can preview 495cb14 at https://pr40369-495cb14.ngbuilds.io/.

AndrewKushnir commented:
@gkalpak thanks for additional feedback, I've applied the necessary changes and also did manual testing that there are no regressions in the way contributors are presented on angular.io, so this PR is ready to go. Thank you.
